RavenDB-12022 :  CountesStorage redesign

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-12022

aviv86 committed:
RavenDB-12022 : introduce BlittableJsonReaderObject.RawBlob

aviv86 committed:
RavenDB-12022 : CountersStorage part

aviv86 committed:
RavenDB-12022 : replication part

aviv86 committed:
RavenDB-12022 : re-implement counters import/export via smuggler
- write counters in batches
- handle reading legacy counters
- apply changes to counters handler

aviv86 committed:
RavenDB-12022 : minor changes in a few tests

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -117,31 +132,11 @@ public IEnumerable<CounterDetail> GetCountersFrom(DocumentsOperationContext cont
-        }
-
-        public long GetNumberOfCountersToProcess(DocumentsOperationContext context, string collection, long afterEtag, out long totalCount)
+       public long GetNumberOfCountersToProcess(DocumentsOperationContext context, string collection, long afterEtag, out long totalCount)
```
GetNumberOfCountersToProcess does ont exactly number of counters, am I right? I think the name should be more accurate here, like get number of documents to process (which also sounds strange)

aviv86 replied:
yes, this was used only in counters ETL,  which doesn't work at all now. when i'll fix ETL part i'll change the name as well. maybe GetNumberOfCounterGroupsToProcess ?

maximburyak replied:
Yes, sounds good

aviv86 replied:
done

maximburyak commented on src/Raven.Client/Documents/Operations/Counters/CountersDetail.cs:
```diff
@@ -46,4 +46,26 @@ public DynamicJsonValue ToJson()
+
+    public class CounterGroupDetail
+    {
+        public string DocumentId { get; set; }
```
Can DocumentId and ChangeVector be LazyStringValue instead of string?

ayende replied:
We probably should, yes

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+
+                    if (data.Modifications != null)
+                    {
+                        data = context.ReadObject(data, null, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
```
I think we should dispose the "old" data here

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+            var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
+            var table = GetCountersTable(context.Transaction.InnerTransaction, collectionName);
+
+            using (DocumentIdWorker.GetLowerIdSliceAndStorageKey(context, documentId, out Slice lowerId, out _))
```
we can probably call it docLowerId

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+        {
+            if (data.TryGet(DbIds, out BlittableJsonReaderArray dbIds))
+            {
+                for (dbIdIndex = 0; dbIdIndex < dbIds.Length; dbIdIndex++)
```
Can DbIDs be a sorted list and a binary search should be performed, or even an object, to do it using the existing blittable mechanisms?

aviv86 replied:
the order of the counter partial-values in a raw blob rely on the order of the the DbIDs, so i don't think a sorted set is a good fit here - inserting a new dbId in the middle of the list will force us to modify all the raw data. 
object can work.

maximburyak replied:
let's go with object then, but first, can you profile that on a multiple dbIndex scenario and see if GetDbIdIndex method is expensive? maybe it's just a preliminary optimization..

ayende replied:
No, this is intentionally a very simple array. In 99.999% of the cases, the number here is going to be < 5, so linear scan is better.

maximburyak replied:
I agree that there could be small difference, but why will the linear scan be better?

ayende replied:
Linear scan for < 5 items is going to always be cheaper than the setup cost of anything else

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+        private static Dictionary<LazyStringValue, int> DbIdToIndex(BlittableJsonReaderArray dbIds)
+        {
+            var count = 0;
+            return new Dictionary<LazyStringValue, int>(dbIds.Items.ToDictionary(x => (LazyStringValue)x, x => count++));
```
It's better to implement the iteration here, instead of using ToDictionary with a lambda call
Also, can we use some kind of dictionary caching/cleaning here as we did with _scopes?

ayende replied:
Why do we need this at all? It would be better to do linear scan each time than pay the allocation / building price of the dictionary

maximburyak replied:
For small sizes (~5) dictionary gives about 10% performance improvement in searching, while we could cache/pool the dictionary objects for memory performance.

aviv86 replied:
removed the dictionary

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+            for (int i = 0; i < sourceCount; i++)
+            {
+                var sourceDbId = (LazyStringValue)sourceDbIds[i];
+                if (dict.TryGetValue(sourceDbId, out var dbIdIndex))
```
Please add caching of the calculated HashCode to LazyStringValue, so this will cost a bit less

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -37,6 +38,12 @@ public unsafe class CountersStorage
+
+        private long _countersCount;
+
+        private readonly List<ByteStringContext<ByteStringMemoryCache>.InternalScope> _scopes = new List<ByteStringContext<ByteStringMemoryCache>.InternalScope>();
```
Please find more meaningfull name for that field - something like _counterModificationMemoryScopes or better (probably).

ayende replied:
I don't like this, you are holding resources here per transaction that are held by an object that outlive it.

maximburyak replied:
the list is "cleare" at the end of the usage - this is only a mean of reusing the list between requests

ayende replied:
I know, but it open us the possibility for missing it later on. In general, we do caching and things like that at the context level

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+                        data = new BlittableJsonReaderObject(existing.Read((int)CountersTable.Data, out int oldSize), oldSize, context);
+                        data = data.Clone(context);
+                        data.TryGet(DbIds, out BlittableJsonReaderArray dbIds);
+                        _scopes.Clear();
```
we should probably clear that field in a finally statement

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+            var maxDbIdIndex = FindMaxDbIdIndex(sourceCount, dict, localDbIds, sourceDbIds);
+            var sourceValues = (CounterValues*)counterValues.Ptr;
+
+            _scopes.Add(context.Allocator.Allocate((maxDbIdIndex + 1) * SizeOfCounterValues, out var newVal));
```
it's probably best to pass _scopes as parameter - it's caching is only an implementation detail

ayende replied:
I really don't like `_scopes`, seems like it should be a value that is passed to the method from caller and then disposes

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+            long value = 0;
+            modified = false;
+            var existingCount = existingCounter.Length / SizeOfCounterValues;
+            var sorted = SortValuesByDbIdIndex(dict, localDbIds, sourceDbIds, source.Length / SizeOfCounterValues);
```
This should be called lazily, so the "sorted" list will be held in the calling method

maximburyak replied:
SortValuesByDbIdIndex's main purpose seems to be  making sure the dict holds all dbIDs, after that, the iteration in line 610 can be performed on the dictionary values, I don't think we need "sorted" at all here, or am I missing something?

aviv86 replied:
no, it's not just to make sure the dict holds all dbIDs.
the main purpose here was to avoid calling `context.Allocator.Allocate()` multiple times and moving the data around.
by sorting the source dbIDs by their local dbId-index, then iterating the sorted array in reverse order, we assure that we have a single call to `context.Allocator.Allocate()` (at most) and that we're not moving the data more than once.

ayende replied:
I don't follow why this is required? But in general we should kill the dictionary and not do any sorting

aviv86 replied:
I don't think we should kill the dictionary.
without a dictionary, doesn't it mean that for each partial value in each counter, 
we will need to do a liner scan on `localDbIds` (and if missing, look in `localDbIds.Modifications` ?) to get the DbId-index?

but yes, maybe sorting can be avoided.
I was think of a case like this one:
- local DbIds = [A, B, C, D]
- local counter 'likes' has partial value of A
- source sends counter 'likes' with partial values of B, C, D
- all of these partial values are missing
- if we don't do any sorting, each iteration will go to `AddPartialValueToExistingCounter()` , allocate `newVal`, copy old data, etc.
- with sorting by local DbId (O(N)), we will go to `AddPartialValueToExistingCounter()` in the first iteration and all the next iterations will modify the data in place.

ayende replied:
Do this, first read the `localDbIds` to a List, then scan this.
Add to this list if needed with new values.

Same logic, simpler behavior. There is no need to do much here, we expect 3 - 5 max.

aviv86 replied:
done

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+        private static int[] SortValuesByDbIdIndex(Dictionary<LazyStringValue, int> dict,
+            BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds, int sourceCount)
+        {
+            var sorted = new int[dict.Count + sourceDbIds.Length];
```
Use an array pool here, probably

aviv86 replied:
I removed this function

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+            return sorted;
+        }
+
+        private static int FindMaxDbIdIndex(int sourceCount, Dictionary<LazyStringValue, int> dict, BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds)
```
I guess that this function can be merged with SortValuesByDbIdIndex, given that it won't need to be sorting anything, but just intializing the dict and modifying the localDbIds

aviv86 replied:
removed this function

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -495,59 +831,29 @@ public IEnumerable<(string ChangeVector, long Value)> GetCounterValues(Documents
         {
```
We can probably abandon the IEnumerable return type for this function and for GetCountersForDocument

aviv86 committed:
RavenDB-12022 : bug fixes + address some PR comments

ppekrol commented on src/Raven.Client/Exceptions/Documents/Counters/CounterOverflowException.cs:
```diff
@@ -16,12 +16,12 @@ public CounterOverflowException(string message, Exception inner) : base(message,
         }
 
-        internal static void ThrowFor(string docId, string counterName, long value, long delta, OverflowException inner)
+        public static void ThrowFor(string docId, string counterName, long value, long delta, OverflowException inner)
```
can be `internal` ?

aviv86 replied:
done

ppekrol commented on src/Raven.Client/Exceptions/Documents/Counters/CounterOverflowException.cs:
```diff
@@ -16,12 +16,12 @@ public CounterOverflowException(string message, Exception inner) : base(message,
         }
 
-        internal static void ThrowFor(string docId, string counterName, OverflowException inner)
+        public static void ThrowFor(string docId, string counterName, OverflowException inner)
```
can be `internal` ?

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+                    // Common case is that we modify the data IN PLACE
+                    // as such, we must copy it before modification
+
+                    data = data.Clone(context);
```
Need to dispose the old data

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+                            return;
+
+                        data = new BlittableJsonReaderObject(existing.Read((int)CountersTable.Data, out int oldSize), oldSize, context);
+                        data = data.Clone(context);
```
Need to dispose the old 'data' ?

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -37,6 +38,12 @@ public unsafe class CountersStorage
+
+        private long _countersCount;
+
+        private readonly List<ByteStringContext<ByteStringMemoryCache>.InternalScope> _scopes = new List<ByteStringContext<ByteStringMemoryCache>.InternalScope>();
```
I do not like that concept. It would be better to add method `RegisterForDispose(IDisposable toDispose)` to DocumentsOperationContext to be absolutely sure that we will dispose everything at the end, even when some exception occured. Thoughts?

ayende replied:
@ppekrol That would be better, yes

maximburyak replied:
We could just use a finally statment, in which we'll call the disposes and .clear. In wost case finalizers will catch that..
In the bottom line, all usages of the _scopes are performed inside the PutCounters function

ppekrol commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,72 +891,131 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
+            data.Modifications = new DynamicJsonValue(data);
+            data.Modifications.Remove(counterName);
+
+            data = context.ReadObject(data, null, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
```
dispose old 'data' ?

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -253,6 +229,218 @@ private static void ThrowInvalidBatchOperationType(CounterOperation operation)
+                _dictionary.Add(id, cg);
+            }
+
+            public void Add(string id, CounterDetail counterDetail, out bool isNew)
```
Rename to `AddLegacy`

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -45,14 +52,22 @@ public unsafe class CountersStorage
-            Collection = 5,
-            TransactionMarker = 6
+            // lower document id, record separator, prefix 
+            DocumentId = 0,
```
CounterKey, instead. This isn't the document id

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -228,61 +203,325 @@ private static (LazyStringValue Doc, LazyStringValue Name) ExtractDocIdAndName(J
+            if (dbIdIndex < existingCount)
+            {
+                exists = true;
+                var counter = (CounterValues*)existingCounter.Ptr + dbIdIndex;
```
```suggestion
                var counter = &((CounterValues*)existingCounter.Ptr)[dbIdIndex];
```

Is that more readable?

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,116 +776,51 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
-                    tvb.Add(context.TransactionMarkerOffset);
-
-                    table.Set(tvb);
+                    if (prop == DbIds)
```
Question, what happens if a user defines a counter named `DbIds` ? Won't it kill us?

ayende replied:
Probably we should define the DbIds property as `@` and avoid a counter of this name

aviv86 replied:
yes `DbIds` already has `@` prefix. we can avoid a counter of this name, but maybe better to wrap all counters in a nested object and call that `@Counters` property ?

ayende replied:
Yes, but let's use shorter name? We care about space.

aviv86 replied:
I'm using `@dbIds` and `@vals`. okay?

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,72 +891,131 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
+            return DeleteCounter(context, doc, collection, name, forceTombstone, lastModifiedTicks);
+        }
+
+        public string DeleteCounter(DocumentsOperationContext context, string documentId, string collection, string counterName, bool forceTombstone = false, long lastModifiedTicks = -1)
```
Do we _need_ a separate tombstone?
Can we mark the property itself on the object as deleted? 
Something like, if counter property is null, it is deleted?

aviv86 replied:
We can do that, but a couple of things to consider here:
- when deleting all counters in a group, we're left with a table entry that's all `null`s
- if a counter was deleted and then created again, how would we deal with that in replication? 
 probably need to keep a deleted etag to know which information is more up do date.

ayende replied:
Not sure, right now we are explicit that delete + modify on counter is something that user need to verify won't happen and might cause the modification to be lost or the delete to be lost

aviv86 replied:
removed CounterTombstones.
As discussed yesterday, deleted entries need to keep a change vector string

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -253,6 +229,218 @@ private static void ThrowInvalidBatchOperationType(CounterOperation operation)
+            public void Add(string id, CounterDetail counterDetail, out bool isNew)
+            {
+                isNew = false;
+                _legacyDictionary = _legacyDictionary ?? new Dictionary<string, Dictionary<string, List<(string ChangeVector, long Value)>>>();
```
Case insensitive on the doc id

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/CountersHandler.cs:
```diff
@@ -253,6 +229,218 @@ private static void ThrowInvalidBatchOperationType(CounterOperation operation)
+
+                foreach (var counter in counters.CounterGroup.Values.GetPropertyNames())
+                {
+                    if (counter.Equals(CountersStorage.DbIds))
```
How can this be?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Smuggler/Documents/StreamDestination.cs:
```diff
@@ -407,34 +407,70 @@ public void WriteIndex(IndexDefinition indexDefinition)
+            public void WriteLegacyCounter(CounterDetail counterDetail)
+            {
+                // Used only in Database Destination 
+                throw new NotImplementedException();
```
NotSupported, and with a proper message

aviv86 replied:
done

aviv86 committed:
RavenDB-12022 : CountersStorage - remove dictionary and avoid sorting + refactoring

aviv86 committed:
RavenDB-12022 : fix PR comments

aviv86 committed:
RavenDB-12022 : remove Counter Tombstones (instead, use an empty blob to indicate that a counter was deleted)

aviv86 committed:
RavenDB-12022 : fix counters part in recovery tool

aviv86 committed:
RavenDB-12022 : fix import/export due to change in counter-group-document structure

aviv86 committed:
RavenDB-12022 : remove ReplicationItemType.CounterTombstone

aviv86 committed:
RavenDB-12022 : initial work on ETL part

aviv86 committed:
RavenDB-12022 :mark  deleted counters entries with a change vector

ayende commented on src/Raven.Client/Documents/Operations/Counters/CounterBatch.cs:
```diff
@@ -146,4 +146,10 @@ public DynamicJsonValue ToJson()
         }
     }
+
+    public class CounterGroup
```
Why do we need this?

aviv86 replied:
removed this. using `CounterGroupDetail` instead

ayende commented on src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:
```diff
@@ -21,6 +21,8 @@ public enum DatabaseItemType
         Unknown = 1 << 12,
 
         Counters = 1 << 13,
```
Mark as `[Obsolete]`

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                {
+                    var dbIdIndex = GetDbIdIndex(context, data);
+
+                    data.TryGet(Values, out BlittableJsonReaderObject counters);
```
Validate that this worked

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+
+                    if (counters.TryGetMember(name, out object existingCounter) == false || 
+                        existingCounter is LazyStringValue ||
+                        !(existingCounter is BlittableJsonReaderObject.RawBlob blob) ||
```
Remove this line

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
-                if (p[sizeOfDocId] == SpecialChars.RecordSeparator)
+                for (dbIdIndex = 0; dbIdIndex < dbIds.Length; dbIdIndex++)
+                {
+                    if (dbIds[dbIdIndex].Equals(context.Environment.Base64Id) == false)
```
Reverse this

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
-                if (p[sizeOfDocId] == SpecialChars.RecordSeparator)
-                    break;
-            }
+                if (dbIdIndex > 0) 
```
Remove this check

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
-        public void PutCounter(DocumentsOperationContext context, string documentId, string collection, string name, long value)
-        {
-            PutCounterImpl(context, documentId, collection, name, null, value);
+                using (AddPartialValueToExistingCounter(context, existingCounter, dbIdIndex, value, newETag))
```
This isn't valid, we are disposing the value while still using `existingCounter`

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                            data = GetData(context, ref existing);
+                            data = data.Clone(context);
+                            data.TryGet(DbIds, out BlittableJsonReaderArray dbIds);
+                            data.TryGet(Values, out BlittableJsonReaderObject localCounters);
```
Need to check these

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+
+                            var localDbIdsList = DbIdsToList(dbIds);
+
+                            sourceData.TryGet(DbIds, out BlittableJsonReaderArray sourceDbIds);
```
check this

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                                    {
+                                        // delete + blob => resolve conflict
+
+                                        using (var localValuesAsCv = RawBlobToChangeVector(context, data, loaclCounterValues, existingChangeVector))
```
You don't need this. Can do the check directly on the change vector by asking to read by id for each of the existing db ids there

ayende replied:
for `loaclCounterValues`, iterate over the internal values, for each non zero etag, go get the db id and then check just that.
You only care if there are all after the delete or no, nothing else

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                                        modified = true;
+                                        changeType = CounterChangeTypes.Delete;
+                                        localCounters.Modifications = localCounters.Modifications ?? new DynamicJsonValue(localCounters);
+                                        // todo add local dbId + etag ? 
```
No need, the original is sufficent.

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                                }
+
+                                if (prop.Value is BlittableJsonReaderObject.RawBlob sourceBlob)
+                                {
```
Extract this to a function that operate over delete cv and counter value.
Should't matter which direction

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+            return localDbIdsList;
+        }
+
+        private long InternalPutCounter(DocumentsOperationContext context, BlittableJsonReaderObject counters, string counterName,
```
`localDbIds` - extract to a struct that contains both the array and the list in one shot. And can apply operations to them

ayende replied:
Should return `bool` here

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+            for (var index = 0; index < sourceCount; index++)
+            {
+                var sourceDbId = (LazyStringValue)sourceDbIds[index];
+                var sourceValue = ((CounterValues*)source.Ptr)[index];
```
Use pointer to avoid copying this

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+            BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds, List<LazyStringValue> localDbIdsList, 
+            BlittableJsonReaderObject.RawBlob existingCounter, BlittableJsonReaderObject.RawBlob source, out bool modified)
+        {
+            long value = 0;
```
Drop this, sum it up in the end which is easier

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                {
+                    var localValuePtr = (CounterValues*)existingCounter.Ptr + localDbIdIndex;
+                    if (localValuePtr->Etag >= sourceValue.Etag ||
+                        sourceDbId.Equals(context.Environment.Base64Id))
```
Remove the `sourceDbId` clause

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                    using (table.Allocate(out TableValueBuilder tvb))
                     {
-                        if (table.ReadByKey(counterKey, out var existing))
+                        if (changeVector != null && table.ReadByKey(counterKey, out var existing))
```
```suggestion
                        if ( table.ReadByKey(counterKey, out var existing))
```

Can't use the CV here

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+            return value;
+        }
+
+        private static int GetOrAddDbIdIndex(BlittableJsonReaderArray localDbIds, List<LazyStringValue> localDbIdsList, LazyStringValue dbId)
```
Move this to dedicated struct

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,219 +809,116 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
-                var etag = _documentsStorage.GenerateNextEtag();
-                var result = ChangeVectorUtils.TryUpdateChangeVector(_documentDatabase.ServerStore.NodeTag, _documentsStorage.Environment.Base64Id, etag, string.Empty);
+                var data = GetData(context, ref existing);
+                data.TryGet(Values, out BlittableJsonReaderObject counters);
```
Paranoia, please

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,219 +809,116 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
         }
 
-        public IEnumerable<string> GetCountersForDocument(DocumentsOperationContext context, string docId)
+        private static BlittableJsonReaderObject GetData(JsonOperationContext context, ref TableValueReader existing)
```
```suggestion
        private static BlittableJsonReaderObject GetCounterValuesData(JsonOperationContext context, ref TableValueReader existing)
```

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,219 +809,116 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
+                if (data.TryGet(Values, out BlittableJsonReaderObject counters) == false ||
+                    counters.TryGetMember(counterName, out object counterValues) == false ||
+                    counterValues is LazyStringValue || // deleted
+                    !(counterValues is BlittableJsonReaderObject.RawBlob counterValuesAsRawBlob))
```
Remove this line

ayende replied:
```suggestion
  
```

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,219 +809,116 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
-        }
-
-        public ByteStringContext.InternalScope GetCounterKey(DocumentsOperationContext context, string documentId, string name, string changeVector, out Slice partialKeySlice)
+        public IEnumerable<(string ChangeVector, long Value)> GetCounterValues(DocumentsOperationContext context, string docId, string counterName)
```
Check where this is used. This should only be returned as a debug info / specific call. 
Shouldn't be in general use

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -375,219 +809,116 @@ public Table GetCountersTable(Transaction tx, CollectionName collection)
+                    data.TryGet(Values, out BlittableJsonReaderObject counters) == false ||
+                    counters.TryGetMember(counterName, out object counterValues) == false ||
+                    counterValues is LazyStringValue || // deleted
+                    !(counterValues is BlittableJsonReaderObject.RawBlob blob))
```
```suggestion
```

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,89 +933,170 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
             {
-                table.DeleteByPrimaryKeyPrefix(keyPrefix);
+                //todo update Count of Counters
+                table.DeleteByKey(lowerId /*, holder => _countersCount -= GetData(context, ref holder.Reader).Count -1*/); 
```
Remove this, we won't track the number of counters for now

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,89 +933,170 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
+                    return null;
+
+                if (counterToDelete is LazyStringValue || // already deleted
+                    !(counterToDelete is BlittableJsonReaderObject.RawBlob blob))
```
```suggestion
```

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,89 +933,170 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
         }
 
-        public string DeleteCounter(DocumentsOperationContext context, Slice key, string collection, long lastModifiedTicks, bool forceTombstone)
+        private void RemoveCounterFromBlittableAndUpdateTable(DocumentsOperationContext context, string counterName, BlittableJsonReaderObject data,
```
See if you can trim the number of params here

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,89 +933,170 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
 
-            var newEtag = _documentsStorage.GenerateNextEtag();
-            _documentsStorage.EnsureLastEtagIsPersisted(context, newEtag);
+        private static LazyStringValue RawBlobToChangeVector(DocumentsOperationContext context, BlittableJsonReaderObject data, BlittableJsonReaderObject.RawBlob counterToDelete, string changeVector,
```
drop this

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -602,89 +933,170 @@ public void DeleteCountersForDocument(DocumentsOperationContext context, string
         }
 
-        private void CreateTombstone(DocumentsOperationContext context, Slice keySlice, string collectionName, long deletedEtag, long lastModifiedTicks, long newEtag, string newChangeVector)
+        private static void AddLocalNodeToChangeVector(DocumentsOperationContext context, StringBuilder sb)
```
drop this

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -797,5 +1209,38 @@ private static SortedSet<string> GetCountersForDocument(BlittableJsonReaderArray
+            for (int i = 0; i < counters.Count; i++)
+            {
+                counters.GetPropertyByIndex(i, ref prop);
+
```
Need to handle tombstones here

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/Handlers/StatsHandler.cs:
```diff
@@ -67,7 +67,7 @@ private void FillDatabaseStatistics(DatabaseStatistics stats, DocumentsOperation
                 stats.SizeOnDisk = size.Data;
                 stats.NumberOfTransactionMergerQueueOperations = Database.TxMerger.NumberOfQueuedOperations;
-                stats.CountOfCounters = Database.DocumentsStorage.CountersStorage.GetNumberOfCounterEntries(context);
+                stats.CountOfCounters = Database.DocumentsStorage.CountersStorage.GetNumberOfCounterEntries();
```
Remove `CountOfCounters`

maximburyak replied:
@ayende can we have the number of "counter entries" here (instead of the percise counters count) this way we'll have an indication of the order of magnitude of counters. It's the same as tracking the attachment count - there could be 1kb attachment and 10mb attachment, we still treat the equally in it's count field

ayende replied:
Fine by me, yes, but needs to be carefully explained.

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -840,21 +840,10 @@ private unsafe void ReadItemsFromSource(ref UnmanagedWriteBuffer writeBuffer, in
+                    item.DocumentSize = sizeOfData;
 
-                    item.LastModifiedTicks = *(long*)ReadExactly(sizeof(long));
+                    ReadExactly(sizeOfData, ref writeBuffer);
```
Need to handle 4.1 -> 4.2 replication counter (error, abort connection)
4.2 -> 4.1 want to replicate (error, abort connection)

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:
```diff
@@ -926,6 +926,58 @@ private unsafe void WriteCounterToServer(DocumentsOperationContext context, Repl
+                        ThrowTooManyChangeVectorEntries(item);
+
+                    var tempBufferPos = 0;
+                    pTemp[tempBufferPos++] = (byte)item.Type;
```
ensure the value is different from the value in 41

aviv86 replied:
done

aviv86 committed:
RavenDB-12022 : avoid tracking the number of counters

aviv86 committed:
RavenDB-12022 : keep node tags along with dbIds, introduce CompareCounterValuesAndDeletedCounter method, fix bug in counter replication

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -312,34 +308,37 @@ public string PutCounter(DocumentsOperationContext context, string documentId, s
                 for (dbIdIndex = 0; dbIdIndex < dbIds.Length; dbIdIndex++)
                 {
-                    if (dbIds[dbIdIndex].Equals(context.Environment.Base64Id) == false)
+                    if (dbIds[dbIdIndex].Equals(_localNodeTagAndDbId) == false)
```
We should hold only the db id here, not the node tag

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -417,7 +412,7 @@ private BlittableJsonReaderObject WriteNewCountersDocument(DocumentsOperationCon
 
                     builder.StartWriteArray();
-                    builder.WriteValue(context.Environment.Base64Id);
+                    builder.WriteValue(_localNodeTagAndDbId);
```
just db id here. The node tag may change over time, and we don't want to mess things up

aviv86 replied:
done

aviv86 committed:
RavenDB-12022 : fixes in ETL and Smuggler due to the change in deleted counters handling

aviv86 committed:
RavenDB-12022 : fix PR comments

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -1051,10 +1034,10 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
-            }
+            }*/
 
             /*            var size = Encoding.UTF8.GetMaxByteCount(sb.Length);
```
let's try making this work

maximburyak commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -1242,5 +1220,74 @@ public static void ConvertFromBlobToNumbers(JsonOperationContext context, Counte
+                bool found = false;
+
+                var current = (CounterValues*)counterValues.Ptr + i;
+                var dbId = dbIds[i].ToString().Substring(2);
```
let's find a way to do that without materialization

aviv86 replied:
done

aviv86 committed:
RavenDB-12022 ; remove node tags from dbIds and deleted counters

aviv86 committed:
RavenDB-12022 : fix PR comments

aviv86 committed:
RavenDB-12022 : counter replication - throw on attempt to replicate counters between v4.2 and v4.1 (both ways)

aviv86 committed:
RavenDB-12022 : remove CounterGroup, use CounterGroupDetail instead

aviv86 committed:
track number of counter groups in ETL progress

aviv86 committed:
RavenDB-12022 : fix PR comments

aviv86 committed:
RavenDB-12022 : counters smuggler fixes
- use a separate context to read items from StreamSource
- Do not dispose data after reading from it StreamSource, dispose it in SmugglerCounterBatchCommand.Dispose()

aviv86 committed:
RavenDB-12022 : minor fixes after rebase

aviv86 committed:
RavenDB-12022 : Schema Upgrade part

aviv86 reviewed (commented):
@arekpalinski please review ETL part

arekpalinski commented:
@aviv86 I'll do it tomorrow morning

arekpalinski commented on test/Tests.Infrastructure/RavenTestBase.cs:
```diff
@@ -603,7 +603,7 @@ protected IDisposable RestoreDatabase(IDocumentStore store, RestoreBackupConfigu
 
             var operation = store.Maintenance.Server.Send(restoreOperation);
-            operation.WaitForCompletion(timeout ?? TimeSpan.FromSeconds(30));
+            operation.WaitForCompletion(timeout ?? TimeSpan.FromSeconds(55530)); //todo revert
```
TODO to revert timeout

ayende commented:
@aviv86 The PR is already pretty big with a lot of comments. Can you create a new clean one so we can review the final state?
